src/main/java/com/commercetools/sync/Syncer.java

```diff
@@ -0,0 +1,106 @@
+package com.commercetools.sync;
+
+import static com.commercetools.sync.commons.utils.CtpQueryUtils.queryAll;
+import static com.commercetools.sync.utils.SphereClientUtils.CTP_SOURCE_CLIENT;
+import static com.commercetools.sync.utils.SphereClientUtils.CTP_TARGET_CLIENT;
+import static com.commercetools.sync.utils.SphereClientUtils.closeCtpClients;
+import static com.commercetools.sync.utils.StatisticsUtils.logStatistics;
+import static com.neovisionaries.i18n.LanguageCode.io;
+import static java.lang.String.format;
+
+import com.commercetools.sync.commons.BaseSync;
+import com.commercetools.sync.commons.BaseSyncOptions;
+import com.commercetools.sync.commons.helpers.BaseSyncStatistics;
+import io.sphere.sdk.models.Resource;
+import io.sphere.sdk.queries.QueryDsl;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base class of the syncer that handles syncing a resource from a source CTP project to a target
+ * CTP project.
+ *
+ * @param <T> The type of the resource (e.g. {@link io.sphere.sdk.products.Product}, {@link
+ *     io.sphere.sdk.categories.Category}, etc..)
+ * @param <S> The type of the resource draft (e.g. {@link io.sphere.sdk.products.ProductDraft},
+ *     {@link io.sphere.sdk.categories.CategoryDraft}, etc..)
+ * @param <U> The type of the sync statistics resulting from the sync process (e.g. {@link
+ *     com.commercetools.sync.products.helpers.ProductSyncStatistics}, {@link
+ *     com.commercetools.sync.categories.helpers.CategorySyncStatistics}, etc..)
+ * @param <V> The type of the sync options used for the sync (e.g. {@link
+ *     com.commercetools.sync.products.ProductSyncOptions}, {@link
+ *     com.commercetools.sync.categories.CategorySyncOptions}, etc..)
+ * @param <C> The type of the query used to query for the source resources (e.g. {@link
+ *     io.sphere.sdk.products.queries.ProductQuery}, {@link
+ *     io.sphere.sdk.categories.queries.CategoryQuery}, etc..)
+ * @param <B> The type of the sync instance used to execute the sync process (e.g. {@link
+ *     com.commercetools.sync.products.ProductSync}, {@link
+ *     com.commercetools.sync.categories.CategorySync}, etc..)
+ */
+public abstract class Syncer<
+    T extends Resource,
+    S,
+    U extends BaseSyncStatistics,
+    V extends BaseSyncOptions<T, S>,
+    C extends QueryDsl<T, C>,
+    B extends BaseSync<S, U, V>> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Syncer.class);
```

Comment:
wdyt about logging in the docker container,  maybe we can add some options to customize logs. `logback.xml` can be used for customizing it. for example, if you run your container in google cloud it can be useful to add a `stack driver appender.`

Reply:
do you know how exactly we can do that? Maybe some resources to read around the topic?

Reply:
Can you please create an issue for it so we address it later?

Reply:
I tried to look for this, It seems the idea of providing different logs for different providers is not a good solution. There are projects like fluentd(`allows you to unify data collection and consumption for a better use and understanding of data.` https://www.fluentd.org/architecture) which stack driver using under the hood. 

So maybe we can discuss what we should log or not but this is not a topic for this conversation. So I'm resolving the conversion as done.

Reply:
as issue for it would be good so we can remember to address it later :)